File: solutions/Python/735.AsteroidCollision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []

        for asteroid in asteroids:
            while stack and asteroid < 0 < stack[-1]:
                top = stack[-1]
                if top < abs(asteroid):
                    stack.pop()
                    continue
                elif top == abs(asteroid):
                    stack.pop()
                break
            else:
                stack.append(asteroid)

        return stack

File: solutions/TypeScript/735.AsteroidCollision.ts
function asteroidCollision(asteroids: number[]): number[] {
    const stack: number[] = [];

    for (const asteroid of asteroids) {
        let shouldPush: boolean = true;

        while (
            stack.length > 0 &&
            asteroid < 0 &&
            stack[stack.length - 1] > 0
        ) {
            const top: number = stack[stack.length - 1];
            const absAsteroid: number = Math.abs(asteroid);

            if (top < absAsteroid) {
                stack.pop();
                continue;
            }

            if (top === absAsteroid) {
                stack.pop();
            }

            shouldPush = false;
            break;
        }

        if (shouldPush) {
            stack.push(asteroid);
        }
    }

    return stack;
}
